### DISCUSSION TITLE
Can't access the shell inside of a container.

### DISCUSSION THREAD

**Discussion by ldvy** [PostgREST/postgrest#2243] [Q&A] [answered]
### Environment
* PostgreSQL version: 11
* PostgREST version: v9.0.0.20220211
* Operating system: Linux

### Description of issue
I need to override the default entrypoint of the postgrest container while deploying to K8s. So something like:
`<other steps> && postgrest`

However, it doesn't work.

I tried to create a dummy container on my computer with the below Dockerfile:
```
FROM postgrest/postgrest:latest

SHELL ["/bin/bash", "-c"]
CMD echo "test" && postgrest
```

I've tried using the /bin/sh and /bin/bash pointers, but in both cases, I got an error saying that /bin/*sh wasn't found.

I wonder if there is any shell present in the PostgREST image.
Please let me know if this is the case.

**Comment by wolfgangwalther** [comment, marked as answer]
No, there is no shell in our docker image. We create a minimal docker image with only the `postgrest` executable. You can find details about this here: https://github.com/PostgREST/postgrest/blob/main/nix/tools/docker/README.md

For some architectures, we don't provide minimal docker images, yet, so those do have a base system including shell installed. However that's nothing we support, but just a side-effect because we can't built the static executable on those architectures, yet.

I assume you'll need to create a config file or something in those steps before executing postgrest. I suggest you do that in an `initContainer` running something like `busybox` or so. You can then mount a volume to that container and put the dynamically created files on that for the postgrest image to consume.

**Reply by ldvy** [reply]
Thank you @wolfgangwalther , I will do it through the initContainer.
And just a quick follow up question:
Is the configuration file name/location static?
If I put the file elsewhere, how would I go around passing its full location to the postgrest binary upon execution?

**Reply by wolfgangwalther** [reply]
https://postgrest.org/en/stable/configuration.html#configuration

**Reply by ldvy** [reply]
@wolfgangwalther ,
As you already mentioned, there is no shell in the image.
In order to pass the configuration file to the postgrest executable, I need to execute the postgrest binary and pass the config file to it as a parameter; so something like:
"postgrest /path/to/file".

The default entrypoint of the image is just a simple execution of the postgrest binary; If I want to specify the config file, I have to override it.
But how can I override it if there is no shell?

Please let me know if there is a different way.

**Reply by ldvy** [reply]
Either way, thank you for your explanation!

**Reply by wolfgangwalther** [reply]
You don't need a shell at all to change the command that the image is run with.

I can do:

```bash
docker run --rm postgrest/postgrest:v9.0.0 postgrest --help
```

This will give me the help output. I can replace `--help` with a path to a configuration file.

For kubernetes, this will be something like:
```yaml
containers:
- image: postgrest/postgrest:v9.0.0
  command: ['postgrest', 'path/to/config']
```

**Reply by FlorinAndrei** [reply]
Having a shell in the image, even from something as minimal as Alpine Linux, is invaluable for troubleshooting. This no-shell policy is far too draconian.

Fortunately, AWS ECS makes it trivial to choose the CPU architecture of a running task, and I was able to get a shell in my Postgrest containers by switching to the ARM image. But users should not be forced to jump through hoops like this just for a shell.

Please consider adding a shell to your Docker images. Alpine Linux is a good starting point, and it's used by many very popular projects, some of them having compute demands far greater than Postgres and therefore ideally suited for heavy optimization of their images - and yet they do provide a shell.

**Reply by wolfgangwalther** [reply]
You can always create your own downstream image that uses whatever shell / distribution you'd like and then just copies postgrest in.

**Reply by steve-chavez** [reply]
> This no-shell policy is far too draconian.

The same restriction also comes with ["distroless" images](https://github.com/GoogleContainerTools/distroless?tab=readme-ov-file#why-should-i-use-distroless-images). The benefits are clear in reduced size + improved security.

A common pattern to debug these type of images is to attach a sidecar container:

- https://matthewsanabria.dev/posts/no-shell-for-you-container/#debugging-with-sidecar-containers
- https://tmp.bearblog.dev/minimal-containers-using-nix/#debugging-a-distroless-image

It seems simple enough (a single command) to attach an alpine container to get shell. I believe what we should for now is to document this practice.

**Reply by FlorinAndrei** [reply]
Except that idea doesn't work everywhere. For example, in AWS ECS you need to change the task definition to stand up a sidecar. That's literally a service redeployment from scratch.

Theory vs. the real world - the ancient dilemma.

**Reply by wolfgangwalther** [reply]
Again, just create your own derived image with your favorite shell / distribution, then copy postgrest in. It's a single static binary, so you won't have any dependency problems. Do this once, then have the shell available forever.
